### DISCUSSION TITLE
GET parameters with the same name (obligatory without [])

### DISCUSSION THREAD

**Discussion by lovrozitnik** [symfony/symfony#49973] [Q&A] [answered]
I am writing an API that has to respond to predefined requests by a third-party software.

The mock request is like this:
http://localhost/files/list?id=one&id=two

Symfony controller can only access the last "id" get parameter from the URL in the controller itself:

```
public function filesListAction(Request $request) {
        dd($request);
}
```

---
The InputBag (from dd($request)) only lists the last "id" parameter value:

```
  +query: Symfony\Component\HttpFoundation\InputBag {#8
    #parameters: array:1 [
      "q_id" => "two"
     ]
  }
```
     
---

I know the default http://localhost/files/list?id[]=one&id[]=two works flawlessly but unfortunately I am not in charge of the third-party software calls. They are set in stone.

How would I approach this problem. I am using Symfony 6.

Thank you in advance!

**Comment by stefangr** [comment]
This is a limitation in PHP.

The query property of the Request object is filled with the values that are available in the global $_GET array.

To be able to read the values from the input without the array notation, you need to manually interpret the QUERY_STRING server variable.

**Reply by lovrozitnik** [reply, marked as answer]
You are correct, I did not know this bubbles up from PHP.

---

What I ended up doing is overwriting $_GET parameters at the start of public/index.php like this:


```
//public/index.php (above autoload)
if (!function_exists('parse_query_string_same_identifier')) {
    function parse_query_string_same_identifier($query_string)
    {
        $query_params = array();

        // Split the query string into an array of key-value pairs
        $query_pairs = explode('&', $query_string);

        // Loop through each key-value pair and join variables with the same name into an array
        foreach ($query_pairs as $query_pair) {
            // Split the key-value pair into its constituent parts
            $pair_parts = explode('=', $query_pair);
            $param_name = urldecode($pair_parts[0]);
            $param_value = isset($pair_parts[1]) ? urldecode($pair_parts[1]) : '';

            // If the parameter name already exists, add the value to the existing array
            if (isset($query_params[$param_name])) {
                // If the existing value is not already an array, convert it to an array
                if (!is_array($query_params[$param_name])) {
                    $query_params[$param_name] = array($query_params[$param_name]);
                }
                $query_params[$param_name][] = $param_value;
            } // If the parameter name is new, add it as a new key-value pair
            else {
                $query_params[$param_name] = $param_value;
            }
        }

        return $query_params;
    }
}
$_GET = parse_query_string_same_identifier($_SERVER['QUERY_STRING']);
```

The code is not yet optimised and/or hack-proof, but it works, since Request::createFromGlobals uses $_GET later in the autoloaded code.

Thanks for the tip!

**Comment by davidoliver-1** [comment]
When using GET parameters in a URL, you can have multiple parameters with the same name. When this happens, the values of the parameters will be combined into an array.

For example, if your URL contains the following parameters:

css
Copy code
example.com/?color=red&color=blue&color=green
The values of the "color" parameter will be combined into an array, like this:

php
Copy code
<?php
$colors = $_GET['color']; // ['red', 'blue', 'green']
?>
You can then loop through the array to perform any necessary actions:

php
Copy code
<?php
$colors = $_GET['color'];
foreach ($colors as $color) {
  // Do something with $color, like printing it to the screen
  echo $color . '<br>';
}
?>
Note that if you have multiple parameters with the same name, the order of the parameters in the URL matters. The values of the parameters will be combined in the order that they appear in the URL.

If you want to avoid having multiple parameters with the same name, you can use different parameter names or use POST parameters instead of GET parameters.
